subsys/usb/device/usb_transfer.c

```diff
@@ -251,7 +251,8 @@ int usb_transfer(uint8_t ep, uint8_t *data, size_t dlen, unsigned int flags,
 
```

Comment:
Please remove "Partially fixes issue https://github.com/zephyrproject-rtos/zephyr/issues/72786" from the commit message.

>Sometimes loosing USB transfers

Why? k_work_submit_to_queue() can fail for various reasons, what happened in your case?

>Tests with echo device

What is "echo device"?

There is still no description of how to reproduce the bug. I doubt that changes in the sample code alone will be enough to reproduce it. Something else has changed in the Zephyr code.

Reply:
> Please remove "Partially fixes issue https://github.com/zephyrproject-rtos/zephyr/issues/72786" from the commit message.

But the 2 commits (formerly one) DO fix the issue. But maybe only adding it to the second commit is fine, too.

Reply:
> Why? k_work_submit_to_queue() can fail for various reasons, what happened in your case?

usb_transfer() returns zero (success) even though the queueing failed for one of various reasons. This causes tx_work_handler() to silently drop data (which is the issue described in #72786)

Reply:
> What is "echo device"?

A board with samples/subsys/usb/cdc_acm firmware (echoing data to the same USB CDC ACM port)

Reply:
> There is still no description of how to reproduce the bug. I doubt that changes in the sample code alone will be enough to reproduce it. Something else has changed in the Zephyr code.

I really don't care so much if you doubt it. The data gets lost, I can see it on wireshark, I can find the reason with a debugger, I can reproduce it (and no, not in 5 minutes, as it's random and was about 1 byte / GB in my case) and the changes fix it. 

Also the description shows the use of functions was just wrong (ignoring return values and states) and should be fixed anyway. With my issue or without it. I am happy to improve my commits as needed and did already add quite some work to this fix but it's time to finish.

Reply:
> > Please remove "Partially fixes issue #72786" from the commit message.
> 
> But the 2 commits (formerly one) DO fix the issue. But maybe only adding it to the second commit is fine, too.

You can add Fixes #72786 to PR description, but "Partially fixes issue #72786" should be removed from commit message.

> > Why? k_work_submit_to_queue() can fail for various reasons, what happened in your case?
> 
> usb_transfer() returns zero (success) even though the queueing failed for one of various reasons. This causes tx_work_handler() to silently drop data (which is the issue described in #72786)

You still do not provide a description for why k_work_submit_to_queue() failed.

> > There is still no description of how to reproduce the bug. I doubt that changes in the sample code alone will be enough to reproduce it. Something else has changed in the Zephyr code.
> 
> I really don't care so much if you doubt it. The data gets lost, I can see it on wireshark, I can find the reason with a debugger, I can reproduce it (and no, not in 5 minutes, as it's random and was about 1 byte / GB in my case) and the changes fix it.

But you should care. Exact steps and environment to reproduce it is still missing tought it was requested few times. The reason why k_work_submit_to_queue() sometimes fails is unknown.

Reply:
> You still do not provide a description for why k_work_submit_to_queue() failed.

Ah, I didn't get this was even the question, I thought it was about what happens after the error response, not before. My bad.

 I don't know what the error code is but it's very likely -EBUSY (as the other ones are for uninit device, and it's working for the overwhelming number of packets). And I don't see why it is relevant to the fix. On the programmer side I understand the need to go deeper into the root cause, but getting busy response from the queue doesn't seem to be something to bother at a rate of less than 1 per billion tries. Is it that unexpected?

At this low error rate finding the issue would take days or weeks without any real win. There can be a new issue to investigate, but for the USB data loss this is the fix and it took already quite some time to find the issues.

Reply:
> I don't know what the error code is but it's very likely -EBUSY (as the other ones are for uninit device, and it's working for the overwhelming number of packets). And I don't see why it is relevant to the fix. On the programmer side I understand the need to go deeper into the root cause, but getting busy response from the queue doesn't seem to be something to bother at a rate of less than 1 per billion tries. 

It is important to find the reason, I still doubt that these changes fix your problem.[Here](https://github.com/zephyrproject-rtos/zephyr/pull/73828#discussion_r1645546441
) you wrote that you can reliably reproduce it. Why is it difficult now?

>Is it that unexpected?

Yes, it is unexpected that k_work_submit_to_queue() fails. Btw, what work_queue is used in your configuration?
 
> At this low error rate finding the issue would take days or weeks without any real win. There can be a new issue to investigate, but for the USB data loss this is the fix and it took already quite some time to find the issues.

This PR cannot be considered to fix #72786 until the reason why k_work_submit failed is clear.